### PR TITLE
Sprint 1: Permissions on a file

### PR DESCRIPTION
**Pull request by danielbalkum**
This PR added an access control list to a file's metadata and the code to reference the requestor of a file against the access control list. The username of a client is now also passed into the system.

### CODE REVIEW

File: rice-namenode/source/main.cc [hunk 1]
```diff
@@ -137,4 +139,14 @@ int main(int argc, char *argv[]) {
       &nncli,
       port);
   server.serve(io_service);
+
+  if (admin) {
```
Comment (ghost) [inline rice-namenode/source/main.cc]: what's the reasoning behind this flag?

File: zookeeper/include/zk_nn_client.h [hunk 2]
```diff
@@ -33,8 +35,12 @@ typedef struct {
   // java.lang.String)
   std::uint64_t access_time;
   std::uint64_t modification_time;
-  char owner[256];  // the client who created the file
-  char group[256];
+  char owner[MAX_USERNAME_LEN];  // the client who created the file
+  char group[MAX_USERNAME_LEN];
+  // TODO(danielbalkum): justify max of 20 readers on a file. Higher? Lower?
```
Comment (ghost) [inline zookeeper/include/zk_nn_client.h]: we can probably get away with less for now but
i'd say just decide now and remove this todo.

File: zookeeper/include/zk_nn_client.h [hunk 3]
```diff
@@ -112,41 +123,67 @@ class ZkNnClient : public ZkClientCommon {
    * @param zk_in shared pointer to a ZKWrapper
    * @return ZkNnClient
    */
-  explicit ZkNnClient(std::shared_ptr<ZKWrapper> zk_in);
+  explicit ZkNnClient(std::shared_ptr<ZKWrapper> zk_in,
```
Comment (ghost) [inline zookeeper/include/zk_nn_client.h]: can you add documentation for the changed method signatures in this file

File: rice-namenode/source/main.cc (hunk 1 above)
Comment (LINKIWI) [reply rice-namenode/source/main.cc]: @heliumj

File: zookeeper/include/zk_nn_client.h (hunk 2 above)
Comment (LINKIWI) [reply zookeeper/include/zk_nn_client.h]: @danielbalkum

File: zookeeper/include/zk_nn_client.h (hunk 3 above)
Comment (LINKIWI) [reply zookeeper/include/zk_nn_client.h]: @yukichul

File: rice-namenode/source/main.cc (hunk 1 above)
Comment (heliumj) [reply rice-namenode/source/main.cc]: We are creating the admin file only for the admin

Comment (heliumj) [reply rice-namenode/source/main.cc]: Just pushed a new commit to fix this

**Review by LINKIWI — APPROVED** [review]
Approving to unblock. Merge after resolving the conflict if build is green

**Review by ghost — APPROVED** [review]

**Pull request merged**